File: tests/queries.test.ts
import { describe, it, expect, beforeEach } from "vitest";
import { EventLog } from "@aetherwing/fcp-core";
import { dispatchQuery } from "../src/queries.js";
import { dispatchOp } from "../src/ops.js";
import { createEmptyConfig, rebuildLabelIndex, createBlock, addBlock, addConnection, generateId } from "../src/model.js";
import type { TerraformConfig, TerraformEvent, Connection } from "../src/types.js";

describe("dispatchQuery", () => {
  let config: TerraformConfig;
  let log: EventLog<TerraformEvent>;

  beforeEach(() => {
    config = createEmptyConfig("Test Infrastructure");
    rebuildLabelIndex(config);
    log = new EventLog<TerraformEvent>();
  });

  // Helper: add blocks via dispatchOp to keep indexes in sync
  function addResource(type: string, label: string, params: Record<string, string> = {}) {
    dispatchOp(
      { verb: "add", positionals: ["resource", type, label], params, selectors: [], raw: "" },
      config, log,
    );
  }

  function addProvider(name: string, params: Record<string, string> = {}) {
    dispatchOp(
      { verb: "add", positionals: ["provider", name], params, selectors: [], raw: "" },
      config, log,
    );
  }

  function addVariable(name: string, params: Record<string, string> = {}) {
    dispatchOp(
      { verb: "add", positionals: ["variable", name], params, selectors: [], raw: "" },
      config, log,
    );
  }

  function addOutput(name: string, params: Record<string, string> = {}) {
    dispatchOp(
      { verb: "add", positionals: ["output", name], params, selectors: [], raw: "" },
      config, log,
    );
  }

  function addData(type: string, label: string, params: Record<string, string> = {}) {
    dispatchOp(
      { verb: "add", positionals: ["data", type, label], params, selectors: [], raw: "" },
      config, log,
    );
  }

  function connectBlocks(src: string, tgt: string) {
    dispatchOp(
      { verb: "connect", positionals: [src, "->", tgt], params: {}, selectors: [], raw: "" },
      config, log,
    );
  }

  // ── MAP ─────────────────────────────────────────────────

  describe("map", () => {
    it("returns config summary", () => {
      addProvider("aws", { region: "us-east-1" });
      addResource("aws_instance", "web", { ami: "ami-123" });
      addResource("aws_s3_bucket", "assets");
      addVariable("env");
      addOutput("web_ip", { value: "aws_instance.web.public_ip" });

      const result = dispatchQuery("map", config, log);
      expect(result).toContain("Test Infrastructure");
      expect(result).toContain("Providers: aws");
      expect(result).toContain("Resources (2)");
      expect(result).toContain("aws_instance x1");
      expect(result).toContain("aws_s3_bucket x1");
      expect(result).toContain("Variables: 1");
      expect(result).toContain("Outputs: 1");
    });

    it("shows connections count when present", () => {
      addResource("aws_instance", "web");
      addResource("aws_s3_bucket", "assets");
      connectBlocks("web", "assets");

      const result = dispatchQuery("map", config, log);
      expect(result).toContain("Connections: 1");
    });

    it("handles empty config", () => {
      const result = dispatchQuery("map", config, log);
      expect(result).toContain("Test Infrastructure");
    });
  });

  // ── LIST ────────────────────────────────────────────────

  describe("list", () => {
    beforeEach(() => {
      addResource("aws_instance", "web");
      addResource("aws_instance", "api");
      addResource("aws_s3_bucket", "assets");
      addVariable("env");
    });

    it("lists all blocks", () => {
      const result = dispatchQuery("list", config, log);
      expect(result).toContain("web");
      expect(result).toContain("api");
      expect(result).toContain("assets");
      expect(result).toContain("env");
    });

    it("filters by @type selector", () => {
      const result = dispatchQuery("list @type:aws_instance", config, log);
      expect(result).toContain("web");
      expect(result).toContain("api");
      expect(result).not.toContain("assets");
      expect(result).not.toContain("env");
    });

    it("filters by @kind selector", () => {
      const result = dispatchQuery("list @kind:variable", config, log);
      expect(result).toContain("env");
      expect(result).not.toContain("web");
    });

    it("filters by @provider selector", () => {
      const result = dispatchQuery("list @provider:aws", config, log);
      expect(result).toContain("web");
      expect(result).toContain("assets");
    });

    it("returns message when no blocks found", () => {
      const empty = createEmptyConfig("Empty");
      rebuildLabelIndex(empty);
      const result = dispatchQuery("list", empty, log);
      expect(result).toContain("No blocks found");
    });
  });

  // ── DESCRIBE ────────────────────────────────────────────

  describe("describe", () => {
    it("shows full block details", () => {
      addResource("aws_instance", "web", { ami: "ami-123", instance_type: "t2.micro" });
      const result = dispatchQuery("describe web", config, log);
      expect(result).toContain("resource");
      expect(result).toContain("aws_instance.web");
      expect(result).toContain("provider: aws");
      expect(result).toContain("attributes:");
      expect(result).toContain("ami");
      expect(result).toContain("ami-123");
      expect(result).toContain("instance_type");
    });

    it("shows tags", () => {
      addResource("aws_instance", "web");
      dispatchOp(
        { verb: "style", positionals: ["web"], params: { tags: "Name=WebServer" }, selectors: [], raw: "" },
        config, log,
      );
      const result = dispatchQuery("describe web", config, log);
      expect(result).toContain("tags:");
      expect(result).toContain("Name");
      expect(result).toContain("WebServer");
    });

    it("shows connections", () => {
      addResource("aws_instance", "web");
      addResource("aws_s3_bucket", "assets");
      connectBlocks("web", "assets");
      const result = dispatchQuery("describe web", config, log);
      expect(result).toContain("connections:");
      expect(result).toContain("assets");
    });

    it("returns error when no label provided", () => {
      const result = dispatchQuery("describe", config, log);
      expect(result).toContain("requires a LABEL");
    });

    it("returns error for non-existent block", () => {
      const result = dispatchQuery("describe nonexistent", config, log);
      expect(result).toContain("not found");
    });
  });

  // ── PLAN ────────────────────────────────────────────────

  describe("plan", () => {
    it("generates valid HCL", () => {
      addProvider("aws", { region: "us-east-1" });
      addResource("aws_instance", "web", { ami: "ami-123", instance_type: "t2.micro" });

      const result = dispatchQuery("plan", config, log);
      expect(result).toContain('provider "aws"');
      expect(result).toContain('region = "us-east-1"');
      expect(result).toContain('resource "aws_instance" "web"');
      expect(result).toContain("ami");
      expect(result).toContain("instance_type");
    });

    it("includes depends_on from connections", () => {
      addResource("aws_instance", "web", { ami: "ami-123" });
      addResource("aws_s3_bucket", "assets");
      connectBlocks("web", "assets");

      const result = dispatchQuery("plan", config, log);
      expect(result).toContain("depends_on = [aws_s3_bucket.assets]");
    });

    it("returns message for empty config", () => {
      const result = dispatchQuery("plan", config, log);
      expect(result).toContain("Empty configuration");
    });
  });

  // ── GRAPH ───────────────────────────────────────────────

  describe("graph", () => {
    it("shows dependency graph", () => {
      addResource("aws_instance", "web");
      addResource("aws_s3_bucket", "assets");
      connectBlocks("web", "assets");

      const result = dispatchQuery("graph", config, log);
      expect(result).toContain("Dependency Graph:");
      expect(result).toContain("web -> assets");
    });

    it("returns message when no connections", () => {
      addResource("aws_instance", "web");
      const result = dispatchQuery("graph", config, log);
      expect(result).toContain("No connections");
    });
  });

  // ── VALIDATE ────────────────────────────────────────────

  describe("validate", () => {
    it("returns valid for well-formed config", () => {
      addProvider("aws", { region: "us-east-1" });
      addResource("aws_instance", "web", { ami: "ami-123" });
      addOutput("web_ip", { value: "aws_instance.web.public_ip" });

      const result = dispatchQuery("validate", config, log);
      expect(result).toContain("valid");
    });

    it("finds missing value on output", () => {
      addOutput("web_ip");
      const result = dispatchQuery("validate", config, log);
      expect(result).toContain("ERROR");
      expect(result).toContain("output");
      expect(result).toContain("value");
    });

    it("finds missing source on module", () => {
      dispatchOp(
        { verb: "add", positionals: ["module", "vpc"], params: {}, selectors: [], raw: "" },
        config, log,
      );
      const result = dispatchQuery("validate", config, log);
      expect(result).toContain("ERROR");
      expect(result).toContain("module");
      expect(result).toContain("source");
    });

    it("warns about resource with no attributes", () => {
      addResource("aws_instance", "web");
      const result = dispatchQuery("validate", config, log);
      expect(result).toContain("WARNING");
      expect(result).toContain("no attributes");
    });
  });

  // ── STATS ───────────────────────────────────────────────

  describe("stats", () => {
    it("returns block counts by kind", () => {
      addProvider("aws");
      addResource("aws_instance", "web");
      addResource("aws_s3_bucket", "assets");
      addVariable("env");
      addOutput("ip", { value: "aws_instance.web.public_ip" });

      const result = dispatchQuery("stats", config, log);
      expect(result).toContain("Total blocks: 5");
      expect(result).toContain("provider: 1");
      expect(result).toContain("resource: 2");
      expect(result).toContain("variable: 1");
      expect(result).toContain("output: 1");
      expect(result).toContain("Connections: 0");
    });
  });

  // ── STATUS ──────────────────────────────────────────────

  describe("status", () => {
    it("returns title and file path", () => {
      const result = dispatchQuery("status", config, log);
      expect(result).toContain("Title: Test Infrastructure");
      expect(result).toContain("File: (unsaved)");
      expect(result).toContain("Blocks: 0");
      expect(result).toContain("Connections: 0");
    });

    it("shows block and connection counts", () => {
      addResource("aws_instance", "web");
      addResource("aws_s3_bucket", "assets");
      connectBlocks("web", "assets");

      const result = dispatchQuery("status", config, log);
      expect(result).toContain("Blocks: 2");
      expect(result).toContain("Connections: 1");
    });
  });

  // ── FIND ────────────────────────────────────────────────

  describe("find", () => {
    beforeEach(() => {
      addResource("aws_instance", "web_server");
      addResource("aws_s3_bucket", "my_assets");
    });

    it("finds blocks by label", () => {
      const result = dispatchQuery("find web", config, log);
      expect(result).toContain("web_server");
    });

    it("finds blocks by type", () => {
      const result = dispatchQuery("find s3", config, log);
      expect(result).toContain("my_assets");
    });

    it("returns message when nothing found", () => {
      const result = dispatchQuery("find nonexistent", config, log);
      expect(result).toContain("No blocks matching");
    });

    it("returns error when no search text", () => {
      const result = dispatchQuery("find", config, log);
      expect(result).toContain("requires search text");
    });
  });

  // ── UNKNOWN QUERY ───────────────────────────────────────

  describe("unknown query", () => {
    it("returns helpful error", () => {
      const result = dispatchQuery("frobnicate", config, log);
      expect(result).toContain('Unknown query');
      expect(result).toContain("frobnicate");
      expect(result).toContain("Available:");
      expect(result).toContain("map");
      expect(result).toContain("plan");
    });
  });
});

File: internal/terraform/queries_test.go
package terraform

import (
	"strings"
	"testing"

	"github.com/os-tack/fcp-terraform/internal/fcpcore"
)

func setupQueryModel() *TerraformModel {
	m := NewModel("test-infra")
	m.AddProvider("aws", map[string]string{"region": "us-east-1"})
	m.AddResource("aws_instance", "web", map[string]string{
		"ami":           "ami-abc123",
		"instance_type": "t3.micro",
	}, nil)
	m.AddResource("aws_vpc", "main", map[string]string{
		"cidr_block": "10.0.0.0/16",
	}, nil)
	m.AddVariable("region", map[string]string{"type": "string", "default": "us-east-1"})
	m.AddOutput("instance_ip", map[string]string{"value": "aws_instance.web.public_ip"})
	m.Connect("web", "main", "networking")
	return m
}

// ── plan ─────────────────────────────────────────────────

func TestQueryPlan_Empty(t *testing.T) {
	m := NewModel("empty")
	result := DispatchQuery("plan", m, nil)
	if !strings.Contains(result, "Empty configuration") {
		t.Errorf("plan on empty model = %q, want empty message", result)
	}
}

func TestQueryPlan_WithBlocks(t *testing.T) {
	m := setupQueryModel()
	result := DispatchQuery("plan", m, nil)
	if !strings.Contains(result, "aws_instance") {
		t.Errorf("plan should contain aws_instance, got:\n%s", result)
	}
	if !strings.Contains(result, "resource") {
		t.Errorf("plan should contain resource keyword, got:\n%s", result)
	}
}

// ── graph ────────────────────────────────────────────────

func TestQueryGraph_Empty(t *testing.T) {
	m := NewModel("no-conns")
	m.AddResource("aws_instance", "lonely", nil, nil)
	result := DispatchQuery("graph", m, nil)
	if !strings.Contains(result, "No connections") {
		t.Errorf("graph with no connections = %q", result)
	}
}

func TestQueryGraph_WithConnections(t *testing.T) {
	m := setupQueryModel()
	result := DispatchQuery("graph", m, nil)
	if !strings.Contains(result, "web") || !strings.Contains(result, "main") {
		t.Errorf("graph should show web -> main, got:\n%s", result)
	}
	if !strings.Contains(result, "networking") {
		t.Errorf("graph should show edge label, got:\n%s", result)
	}
}

// ── describe ─────────────────────────────────────────────

func TestQueryDescribe_NotFound(t *testing.T) {
	m := setupQueryModel()
	result := DispatchQuery("describe ghost", m, nil)
	if !strings.Contains(result, "not found") {
		t.Errorf("describe ghost = %q, want not found", result)
	}
}

func TestQueryDescribe_Resource(t *testing.T) {
	m := setupQueryModel()
	result := DispatchQuery("describe web", m, nil)
	if !strings.Contains(result, "aws_instance.web") {
		t.Errorf("describe web should show qualified name, got:\n%s", result)
	}
	if !strings.Contains(result, "ami") {
		t.Errorf("describe web should show attributes, got:\n%s", result)
	}
}

func TestQueryDescribe_Connections(t *testing.T) {
	m := setupQueryModel()
	result := DispatchQuery("describe web", m, nil)
	if !strings.Contains(result, "connections") {
		t.Errorf("describe web should show connections, got:\n%s", result)
	}
	if !strings.Contains(result, "main") {
		t.Errorf("describe web should show connection to main, got:\n%s", result)
	}
}

func TestQueryDescribe_NoLabel(t *testing.T) {
	m := setupQueryModel()
	result := DispatchQuery("describe", m, nil)
	if !strings.Contains(result, "requires a LABEL") {
		t.Errorf("describe without label = %q", result)
	}
}

// ── stats ────────────────────────────────────────────────

func TestQueryStats(t *testing.T) {
	m := setupQueryModel()
	result := DispatchQuery("stats", m, nil)
	if !strings.Contains(result, "Total blocks: 5") {
		t.Errorf("stats should show 5 blocks, got:\n%s", result)
	}
	if !strings.Contains(result, "resource: 2") {
		t.Errorf("stats should show 2 resources, got:\n%s", result)
	}
	if !strings.Contains(result, "Connections: 1") {
		t.Errorf("stats should show 1 connection, got:\n%s", result)
	}
}

// ── map ──────────────────────────────────────────────────

func TestQueryMap(t *testing.T) {
	m := setupQueryModel()
	result := DispatchQuery("map", m, nil)
	if !strings.Contains(result, "test-infra") {
		t.Errorf("map should contain title, got:\n%s", result)
	}
	if !strings.Contains(result, "Providers: aws") {
		t.Errorf("map should list providers, got:\n%s", result)
	}
	if !strings.Contains(result, "Resources (2)") {
		t.Errorf("map should show resource count, got:\n%s", result)
	}
	if !strings.Contains(result, "Variables: 1") {
		t.Errorf("map should show variable count, got:\n%s", result)
	}
}

// ── status ───────────────────────────────────────────────

func TestQueryStatus(t *testing.T) {
	m := setupQueryModel()
	result := DispatchQuery("status", m, nil)
	if !strings.Contains(result, "Title: test-infra") {
		t.Errorf("status should show title, got:\n%s", result)
	}
	if !strings.Contains(result, "(unsaved)") {
		t.Errorf("status should show unsaved, got:\n%s", result)
	}
}

func TestQueryStatus_WithPath(t *testing.T) {
	m := setupQueryModel()
	m.FilePath = "/tmp/main.tf"
	result := DispatchQuery("status", m, nil)
	if !strings.Contains(result, "/tmp/main.tf") {
		t.Errorf("status should show file path, got:\n%s", result)
	}
}

// ── history ──────────────────────────────────────────────

func TestQueryHistory_NoLog(t *testing.T) {
	m := setupQueryModel()
	result := DispatchQuery("history", m, nil)
	if !strings.Contains(result, "No event log") {
		t.Errorf("history with nil log = %q", result)
	}
}

func TestQueryHistory_WithEvents(t *testing.T) {
	m := setupQueryModel()
	log := fcpcore.NewEventLog()
	log.Append(&SnapshotEvent{Summary: "add resource aws_instance web"})
	log.Append(&SnapshotEvent{Summary: "set web ami:ami-xyz"})

	result := DispatchQuery("history", m, log)
	if !strings.Contains(result, "1. add resource") {
		t.Errorf("history should show first event, got:\n%s", result)
	}
	if !strings.Contains(result, "2. set web") {
		t.Errorf("history should show second event, got:\n%s", result)
	}
}

func TestQueryHistory_WithCount(t *testing.T) {
	m := setupQueryModel()
	log := fcpcore.NewEventLog()
	log.Append(&SnapshotEvent{Summary: "event 1"})
	log.Append(&SnapshotEvent{Summary: "event 2"})
	log.Append(&SnapshotEvent{Summary: "event 3"})

	result := DispatchQuery("history 2", m, log)
	// Recent(2) returns last 2 events
	if !strings.Contains(result, "event 2") || !strings.Contains(result, "event 3") {
		t.Errorf("history 2 should show last 2 events, got:\n%s", result)
	}
}

// ── list ─────────────────────────────────────────────────

func TestQueryList_All(t *testing.T) {
	m := setupQueryModel()
	result := DispatchQuery("list", m, nil)
	if !strings.Contains(result, "aws_instance.web") {
		t.Errorf("list should show web, got:\n%s", result)
	}
	if !strings.Contains(result, "aws_vpc.main") {
		t.Errorf("list should show main, got:\n%s", result)
	}
}

func TestQueryList_WithSelector(t *testing.T) {
	m := setupQueryModel()
	result := DispatchQuery("list @type:aws_instance", m, nil)
	if !strings.Contains(result, "web") {
		t.Errorf("list @type should show web, got:\n%s", result)
	}
	if strings.Contains(result, "main") {
		t.Errorf("list @type:aws_instance should NOT show vpc main, got:\n%s", result)
	}
}

func TestQueryList_Empty(t *testing.T) {
	m := NewModel("empty")
	result := DispatchQuery("list", m, nil)
	if !strings.Contains(result, "No blocks") {
		t.Errorf("list on empty = %q", result)
	}
}

// ── find ─────────────────────────────────────────────────

func TestQueryFind(t *testing.T) {
	m := setupQueryModel()
	result := DispatchQuery("find web", m, nil)
	if !strings.Contains(result, "aws_instance.web") {
		t.Errorf("find web = %q", result)
	}
}

func TestQueryFind_NoMatch(t *testing.T) {
	m := setupQueryModel()
	result := DispatchQuery("find unicorn", m, nil)
	if !strings.Contains(result, "No blocks matching") {
		t.Errorf("find unicorn = %q", result)
	}
}

func TestQueryFind_ByType(t *testing.T) {
	m := setupQueryModel()
	result := DispatchQuery("find vpc", m, nil)
	if !strings.Contains(result, "aws_vpc.main") {
		t.Errorf("find vpc should match type, got:\n%s", result)
	}
}

// ── unknown ──────────────────────────────────────────────

func TestQueryUnknown(t *testing.T) {
	m := setupQueryModel()
	result := DispatchQuery("bogus", m, nil)
	if !strings.Contains(result, "Unknown query") {
		t.Errorf("unknown query = %q", result)
	}
}

func TestQueryEmpty(t *testing.T) {
	m := setupQueryModel()
	result := DispatchQuery("", m, nil)
	if !strings.Contains(result, "Query required") {
		t.Errorf("empty query = %q", result)
	}
}
